witness/src/Dscp/Snowdrop/Storage/Avlp.hs

```diff
@@ -0,0 +1,236 @@
+
+      runWrapper rdb $ AVL.overwrite (acaMap accum)
+      return proof
+
+mkAccessActions
+  :: forall h k v m
+  .  ( IdSumPrefixed k
+     , AVLBase h k v m
+     )
+  => RDB.RocksDB
+  -> TVar (Set.Set AVL.Revision)
+  -> DbAccessActions (AVLChgAccum h k v) k v m
+mkAccessActions rdb lookups = DbAccessActions
+    { daaGetter
+    , daaModifyAccum
+    , daaIter
+    }
+  where
+    storeNodeset :: Set.Set AVL.Revision -> m ()
+    storeNodeset nodeset = do
+        atomically $ STM.modifyTVar lookups (<> nodeset)
+
+    daaGetter :: AVLChgAccum h k v -> StateR k -> m (StateP k v)
+    daaGetter accum req = do
+        let
+          getOne newAccum key = do
+            ((val, proof), _) <- runWrapper rdb $ AVL.lookup key (acaMap newAccum)
+            storeNodeset proof
+            return $ maybe Map.empty (Map.singleton key) val
+
+        ss <- mapM (getOne accum) (Set.toList req)
+        return $ Map.unions ss
+
+    daaIter
+        :: forall b
+        .  AVLChgAccum h k v
+        -> Prefix
+        -> b
+        -> ((k, v) -> b -> b)
+        -> m b
+    daaIter accum prefix start folder = do
+        (result, nodeset) <- runWrapper rdb $ AVL.foldIf (\x -> idSumPrefix x == prefix, start, folder, id) (acaMap accum)
+        storeNodeset nodeset
+        return result
+
+    daaModifyAccum
+        :: AVLChgAccum h k v
+        -> ChgAccumModifier k v
+        -> m (Either (CSMappendException k) (AVLChgAccum h k v, Undo k v))
+    daaModifyAccum oldAccum changes = do
+        case changes of
+          CAMChange cs -> do
+            applyChangeSet cs oldAccum
+
+          CAMRevert (Undo cs _wat) -> do
+            applyChangeSet cs oldAccum
+
+      where
+        applyChangeSet
+            :: ChangeSet k v
+            -> AVLChgAccum h k v
+            -> m (Either (CSMappendException k) (AVLChgAccum h k v, Undo k v))
+        applyChangeSet (changeSetToList -> csList) accum = do
+            undo <- mapM (computeUndo accum) csList
+            res  <- foldM apply accum csList
+
+            let undoCS = ChangeSet (Map.fromList undo)
+
+            return (Right (res, Undo undoCS BS.empty))
+          `catch` \(e :: CSMappendException k) ->
+            return (Left e)
+
+        apply :: AVLChgAccum h k v -> (k, ValueOp v)  -> m (AVLChgAccum h k v)
+        apply accum (k, dv) = do
+            ((mres, _), _) <- runWrapper rdb $ AVL.lookup k (acaMap accum)
+            case (dv, mres) of
+              (NotExisted, _)      -> return accum
+              (New thing, Nothing) -> accum `withAccum` AVL.insert k thing
+              (Upd thing, Just _)  -> accum `withAccum` AVL.insert k thing
+              (Rem,       Just _)  -> accum `withAccum` AVL.delete k
+              _                    -> throwM $ CSMappendException k
+          `catch` \(_ :: AVL.NotFound h) ->
+            throwM $ CSMappendException k
+
+
+        computeUndo :: AVLChgAccum h k v -> (k, ValueOp v) -> m (k, ValueOp v)
```

Comment:
For AVL+ it's generally wrong to define `Undo := ChangeSet` as operation of applying changes and rolling them back is not equal to identity as rollback will produce not the same root hash as was before application of original changes.

Reply:
I agree here. However, since we now `AVL.overwrite` storage (as was suggested by you), I see no way we will get to same root hash we were before, because we wipe old state from store before saving new one.

Even if remove `AVL.Revision`s and make zipper rehash nodes each time we go `up` from them, I cannot guaratee that `AVL.delete k1` would reverse state to one before `AVL.insert k1` - that is because rebalance is not perfect.

It is only possible if I add `AVL.append` method (which doesn't perform any erasure), will use it instead of `AVL.overwrite` and will perform undo via `AVL.assignRoot oldRoot`.

So, we're back at square 1. Undo is only possible on truly persistent storage.

Reply:
We can overwrite on Genesis blocks, btw. By completely wiping storage and writing genesis state into it.